File: android/src/main/kotlin/dev/doclens/doclens/TextRecognizer.kt
package dev.doclens.doclens

import android.graphics.BitmapFactory
import android.graphics.Rect
import com.google.android.gms.tasks.Tasks
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.text.Text
import com.google.mlkit.vision.text.TextRecognition
import com.google.mlkit.vision.text.latin.TextRecognizerOptions
import java.util.concurrent.TimeUnit

/**
 * On-device OCR over a document image, returning the full recognised text plus
 * per-block / per-line bounding boxes and confidence.
 *
 * Uses ML Kit's on-device Latin text recognizer, delivered by Google Play
 * services (`play-services-mlkit-text-recognition`) — exactly like the
 * OS-native document scanner and `AutoOrientation.auto`. No model is bundled in
 * the app; it is downloaded on demand the first time it is used.
 */
object TextRecognizer {
    private val recognizer by lazy {
        TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS)
    }

    /**
     * Recognise text in the image at [path]. The returned map matches the Dart
     * `OcrResult.fromMap` contract:
     * `{ text, imageSize:[w,h], blocks:[ { text, boundingBox:[l,t,w,h],
     *    recognizedLanguage, lines:[ { text, boundingBox, confidence } ] } ] }`.
     *
     * MUST be called off the main thread — it blocks on the recognizer.
     */
    fun recognize(path: String): Map<String, Any?> {
        val raw = BitmapFactory.decodeFile(path)
            ?: throw ScannerException.CaptureFailed("Decode failed: $path")
        // Bake EXIF orientation into the pixels so the bounding boxes are in the
        // same upright pixel space the caller is displaying.
        val bitmap = ExifRotator.rotated(raw, path)
        try {
            val text: Text = try {
                Tasks.await(
                    recognizer.process(InputImage.fromBitmap(bitmap, 0)),
                    10, TimeUnit.SECONDS,
                )
            } catch (_: Exception) {
                // Model still downloading on demand, or recognition failed —
                // return an empty result rather than surfacing an error.
                return emptyResult(bitmap.width, bitmap.height)
            }

            val blocks = ArrayList<Map<String, Any?>>(text.textBlocks.size)
            for (block in text.textBlocks) {
                val lines = ArrayList<Map<String, Any?>>(block.lines.size)
                for (line in block.lines) {
                    lines.add(
                        mapOf(
                            "text" to line.text,
                            "boundingBox" to rectToList(line.boundingBox),
                            "confidence" to confidenceOf(line.confidence),
                        ),
                    )
                }
                blocks.add(
                    mapOf(
                        "text" to block.text,
                        "boundingBox" to rectToList(block.boundingBox),
                        "recognizedLanguage" to block.recognizedLanguage
                            .ifEmpty { null },
                        "lines" to lines,
                    ),
                )
            }

            return mapOf(
                "text" to text.text,
                "imageSize" to listOf(bitmap.width.toDouble(), bitmap.height.toDouble()),
                "blocks" to blocks,
            )
        } finally {
            if (bitmap !== raw) bitmap.recycle()
            raw.recycle()
        }
    }

    private fun emptyResult(width: Int, height: Int): Map<String, Any?> = mapOf(
        "text" to "",
        "imageSize" to listOf(width.toDouble(), height.toDouble()),
        "blocks" to emptyList<Map<String, Any?>>(),
    )

    /** `[left, top, width, height]` in image pixel coordinates, or null. */
    private fun rectToList(r: Rect?): List<Double>? {
        if (r == null) return null
        return listOf(
            r.left.toDouble(),
            r.top.toDouble(),
            r.width().toDouble(),
            r.height().toDouble(),
        )
    }

    /** `confidence` can be null or NaN on some models — drop it then. */
    private fun confidenceOf(conf: Float?): Double? =
        if (conf == null || conf.isNaN()) null else conf.toDouble()
}

File: ios/doclens/Sources/doclens/TextRecognizer.swift
import CoreGraphics
import UIKit
import Vision

/// On-device OCR over a document image, returning the full recognised text
/// plus per-line bounding boxes and confidence.
///
/// Uses Apple Vision's `VNRecognizeTextRequest` (iOS 13+) — part of the OS, no
/// bundled model. Run with the `.accurate` recognition level and language
/// correction on for the best transcription (this is a one-shot call on a
/// captured still, not the live path, so accuracy is worth the cost).
///
/// Vision has no notion of a paragraph "block", so each recognised line is
/// emitted as its own block containing a single line — keeping the cross-
/// platform [OcrResult] shape uniform with Android's ML Kit (blocks → lines).
enum TextRecognizer {
    /// Recognise text in the image at [path]. The returned map matches the
    /// Dart `OcrResult.fromMap` contract:
    /// `{ text, imageSize:[w,h], blocks:[ { text, boundingBox:[l,t,w,h],
    ///    recognizedLanguage, lines:[ { text, boundingBox, confidence } ] } ] }`.
    static func recognize(path: String) throws -> [String: Any] {
        guard let img = UIImage(contentsOfFile: path),
              let cg = img.uprightCGImage() else {
            throw ImageWarperError.ioError("Cannot read image at \(path)")
        }

        let width = CGFloat(cg.width)
        let height = CGFloat(cg.height)

        let request = VNRecognizeTextRequest()
        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = true

        let handler = VNImageRequestHandler(cgImage: cg, options: [:])
        try handler.perform([request])

        var blocks: [[String: Any]] = []
        var fullLines: [String] = []

        for observation in request.results ?? [] {
            guard let candidate = observation.topCandidates(1).first else { continue }
            let string = candidate.string
            if string.isEmpty { continue }

            // Vision boxes are normalised with a bottom-left origin and y-up.
            // Convert to top-left-origin pixel coordinates so the boxes match
            // the image the caller is displaying.
            let bb = observation.boundingBox
            let box: [Double] = [
                Double(bb.minX * width),
                Double((1 - bb.maxY) * height),
                Double(bb.width * width),
                Double(bb.height * height),
            ]

            let line: [String: Any] = [
                "text": string,
                "boundingBox": box,
                "confidence": Double(candidate.confidence),
            ]
            blocks.append([
                "text": string,
                "boundingBox": box,
                "lines": [line],
            ])
            fullLines.append(string)
        }

        return [
            "text": fullLines.joined(separator: "\n"),
            "imageSize": [Double(width), Double(height)],
            "blocks": blocks,
        ]
    }
}
